### PR TITLE
Add `files` option

### PR DESCRIPTION
**Pull request by kevva**
Probably room for some improvement here and there, but it's functioning at least :). This is so we can use this in https://github.com/sindresorhus/globby/issues/33.

@sindresorhus @SamVerschueren @schnittstabil

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -4,27 +4,51 @@ const arrify = require('arrify');
 const isDirectory = require('is-directory');
 const pify = require('pify');
 
-const getGlob = (fp, ext) => path.join(fp, '**', ext || '');
-const getExt = ext => Array.isArray(ext) ? `*.{${ext.join(',')}}` : `*.${ext}`;
-const getPath = fp => fp[0] === '!' ? fp.slice(1) : fp;
+const isDirectoryP = pify(isDirectory);
```
Comment (sindresorhus) [inline index.js]: Instead of using `pify` here, just use https://github.com/sindresorhus/path-type instead

File: index.js [hunk 2]
```diff
@@ -4,27 +4,51 @@ const arrify = require('arrify');
...
 	}
 
-	return input.map(x => isDirectory.sync(getPath(x)) ? getGlob(x, opts.ext) : x);
+	return files.map(x => path.join(dir, x));
+};
+
+module.exports = (input, opts) => {
+	return Promise.all(arrify(input).map(x => isDirectoryP(getPath(x))
+		.then(isDir => isDir ? getGlob(x, opts) : x)))
+		.then(globs => globs.reduce((a, b) => a.concat(b), []));
```
Comment (sindresorhus) [inline index.js]: `[].concat.apply([], globs);` instead?

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
How does the `files` option handle filenames with a dot? Like `webpack.config` where the full name is `webpack.config.js`.

File: index.js (hunk 1 above)
Comment (kevva) [reply index.js]: Used `is-directory` because it doesn't `throw` when the path doesn't exist. Could handle it here, but I thought it was cleaner not to.

Comment (sindresorhus) [reply index.js]: `path-type` doesn't throw either, it just returns a boolean.

Comment (kevva) [reply index.js]: Yes, it does. You aren't handling `ENOENT` anywhere. Tests are failing when using it:

```
convert directories to glob - async

  Rejected promise returned by test. Reason:

  Error {
    code: 'ENOENT',
    errno: -2,
    path: 'foo/**',
    syscall: 'stat',
    message: 'ENOENT: no such file or directory, stat \'foo/**\'',
  }
```

If can do a PR if you want.

Comment (sindresorhus) [reply index.js]: Oh. That is definitely not the intention. PR would be lovely :)

**Review by sindresorhus — APPROVED** [review]

**Comment by kevva** [comment]
> How does the `files` option handle filenames with a dot? Like `webpack.config` where the full name is `webpack.config.js`.

It doesn't atm. Any ideas?

**Comment by sindresorhus** [comment]
We have to decide whether the extension is after the first or last dot.

**Comment by kevva** [comment]
The last dot might be a better default since it'll cover most cases except `.tar.gz` etc. Maybe it should be configurable?

**Comment by sindresorhus** [comment]
I would go with the last dot too. I don't think it should be configurable.

**Comment by kevva** [comment]
Alright. What's your thoughts on https://github.com/sindresorhus/globby/pull/46#issuecomment-312863683v to increase performance?

> I'm wondering if we maybe should check if it's a glob using a regex before checking if it's a directory in the `dir-glob` module?

**Comment by sindresorhus** [comment]
I don't think it's worth it. You don't usually have a lot of glob patterns, and few fs.stat checks are pretty cheap relatively to everything else involved in globbing. Using a regex will have ambiguous cases. The only way to be totally sure is to actually check the disk.

**Comment by kevva** [comment]
Alright, then I think this is good to 🚢 ?

**Comment by sindresorhus** [comment]
👍

**Pull request merged**